lib/auth/join_azure.go

```diff
@@ -0,0 +1,354 @@
+		var unverifiedClaims accessTokenClaims
+		if err := token.UnsafeClaimsWithoutVerification(&unverifiedClaims); err != nil {
+			return nil, trace.Wrap(err)
+		}
+		provider, err := oidc.NewProvider(ctx, fmt.Sprintf("https://sts.windows.net/%v/", unverifiedClaims.TenantID))
+		if err != nil {
+			return nil, trace.Wrap(err)
+		}
+		verifiedToken, err := provider.Verifier(cfg).Verify(ctx, rawIDToken)
+		if err != nil {
+			return nil, trace.Wrap(err)
+		}
+		var tokenClaims accessTokenClaims
+		if err := verifiedToken.Claims(&tokenClaims); err != nil {
+			return nil, trace.Wrap(err)
+		}
+		return &tokenClaims, nil
+	}
+}
+
+func (cfg *azureRegisterConfig) CheckAndSetDefaults(ctx context.Context) error {
+	if cfg.clock == nil {
+		cfg.clock = clockwork.NewRealClock()
+	}
+	if cfg.verify == nil {
+		oidcConfig := &oidc.Config{
+			ClientID: azureAccessTokenAudience,
+			Now:      cfg.clock.Now,
+		}
+		cfg.verify = azureVerifyFuncFromOIDCVerifier(oidcConfig)
+	}
+
+	if cfg.certs == nil {
+		certs, err := getAzureCerts()
+		if err != nil {
+			return trace.Wrap(err)
+		}
+		cfg.certs = certs
+	}
+	return nil
+}
+
+type azureRegisterOption func(cfg *azureRegisterConfig)
+
+// parseAndVeryAttestedData verifies that an attested data document was signed
+// by Azure. If verification is successful, it returns the ID of the VM that
+// produced the document.
+func parseAndVerifyAttestedData(adBytes []byte, challenge string, certs []*x509.Certificate) (string, error) {
+	var signedAD signedAttestedData
+	if err := utils.FastUnmarshal(adBytes, &signedAD); err != nil {
+		return "", trace.Wrap(err)
+	}
+	if signedAD.Encoding != "pkcs7" {
+		return "", trace.AccessDenied("unsupported signature type: %v", signedAD.Encoding)
+	}
+
+	sigPEM := "-----BEGIN PKCS7-----\n" + signedAD.Signature + "\n-----END PKCS7-----"
+	sigBER, _ := pem.Decode([]byte(sigPEM))
+	if sigBER == nil {
+		return "", trace.AccessDenied("unable to decode attested data document")
+	}
+
+	p7, err := pkcs7.Parse(sigBER.Bytes)
+	if err != nil {
+		return "", trace.Wrap(err)
+	}
+	fixAzureSigningAlgorithm(p7)
+
+	pool := x509.NewCertPool()
+	for _, cert := range certs {
+		pool.AddCert(cert)
+	}
+
+	if err := p7.VerifyWithChain(pool); err != nil {
+		return "", trace.Wrap(err)
+	}
+
+	var ad attestedData
+	if err := utils.FastUnmarshal(p7.Content, &ad); err != nil {
+		return "", trace.Wrap(err)
+	}
+
+	if ad.Nonce != challenge {
+		return "", trace.AccessDenied("challenge is missing or does not match")
+	}
+
+	return ad.ID, nil
+}
+
+// verifyVMIdentity verifies that the provided access token came from the
+// correct Azure VM.
+func verifyVMIdentity(ctx context.Context, cfg *azureRegisterConfig, accessToken, vmID string, requestStart time.Time) (*azure.VirtualMachine, error) {
+	tokenClaims, err := cfg.verify(ctx, accessToken)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	expectedIssuer := fmt.Sprintf("https://sts.windows.net/%v/", tokenClaims.TenantID)
```

Comment:
Consider using url.JoinPath here, which will do some extra cleaning and validation.